### ISSUE TITLE
Low GPU utilization in google colab

### ISSUE THREAD

**Issue by berlaychen** [ultralytics/yolov5#3012]
Hello, I'm facing the problem that recently training on google colab, wandb reported that GPU utilization only around 25%
![image](https://user-images.githubusercontent.com/51433626/116806665-35ef8880-ab61-11eb-9154-e96fa1abedb6.png)
A weeks ago it has reached at 60% but now it didn't. Training speed is much lower now, before this can do 75 epoches in an hour but now only ~40 epoches.
have tried in different accounts but also gave me same result, also tried put data in colab local disk but doesn't make good results.
Is there anything I missed? thanks for helping

**Comment by berlaychen** [comment]
Hmmm
![image](https://user-images.githubusercontent.com/51433626/116811622-df904300-ab7c-11eb-832b-8397a7abb69d.png)
Switching to kaggle, and it works like a charm.
speed up at 6 times to colab, really don't know why colab gave me slow performance.
Before figuring out the problem, I'll temporary using kaggle instead, though it's harder to use compare to colab.

**Comment by glenn-jocher** [comment]
@nk950357 results will vary by hardware. High GPU utilization requires an effective dataloading technique, and Colab only assigns 2 vCPUs per user on their free plan.

**Comment by berlaychen** [comment]
> @nk950357 results will vary by hardware. High GPU utilization requires an effective dataloading technique, and Colab only assigns 2 vCPUs per user on their free plan.

so if im in free plan, then i won't get better performance, right?

**Comment by glenn-jocher** [comment]
@nk950357 you can --cache your dataset if your hardware has sufficient RAM:
```
python train.py --cache
```

**Comment by berlaychen** [comment]
> @nk950357 you can --cache your dataset if your hardware has sufficient RAM:
> 
> ```
> python train.py --cache
> ```

Gotcha. will try this way after current training done. 
And also i want to ask is my training in a good result?
![image](https://user-images.githubusercontent.com/51433626/116824493-22720b00-abbd-11eb-8454-d99bef8a70bd.png)
![image](https://user-images.githubusercontent.com/51433626/116824502-2aca4600-abbd-11eb-824d-15f55cd95805.png)
![image](https://user-images.githubusercontent.com/51433626/116824505-2ef66380-abbd-11eb-8dfd-367b8f4807c6.png)
![image](https://user-images.githubusercontent.com/51433626/116824510-3453ae00-abbd-11eb-8d04-78871c2d313c.png)
![image](https://user-images.githubusercontent.com/51433626/116824514-39186200-abbd-11eb-8657-e34ea06b9103.png)

**Comment by glenn-jocher** [comment]
@nk950357 👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by berlaychen** [comment]
> @nk950357 👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.
> 
> If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.
> 
> We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.


![image](https://user-images.githubusercontent.com/51433626/116839010-2de82500-ac03-11eb-979d-7d944d90507f.png)
Ahhhh can confirm adding --cache in colab makes gpu run in high utilization! thanks!

and below is my training data:
<pre> 1349/1349     14.6G   0.01645   0.01472         0   0.03117         9       640: 100%|██████████| 15/15 [00:16<00:00,  1.10s/it]
               Class      Images      Labels           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:03<00:00,  3.46s/it]
                 all          87         186       0.869       0.892       0.876       0.429</pre>

PR_Curve:
![image](https://user-images.githubusercontent.com/51433626/116839296-6ccaaa80-ac04-11eb-92d6-0144996c442c.png)

Confusion matrix:
![image](https://user-images.githubusercontent.com/51433626/116839306-7bb15d00-ac04-11eb-828a-74949451e7c0.png)

Does training mosaics like this?
![image](https://user-images.githubusercontent.com/51433626/116839561-62f57700-ac05-11eb-98c7-4f975210afa4.png)

And is test result like this?
![image](https://user-images.githubusercontent.com/51433626/116839585-7dc7eb80-ac05-11eb-8e21-6cb57f417803.png)

label.png
![image](https://user-images.githubusercontent.com/51433626/116839847-60475180-ac06-11eb-8324-02bb8a8ba521.png)
![image](https://user-images.githubusercontent.com/51433626/116839855-69382300-ac06-11eb-96bd-d97bc3c18ffa.png)
![image](https://user-images.githubusercontent.com/51433626/116839860-6f2e0400-ac06-11eb-8af8-0c5ffe0d0260.png)

sorry for my bad english understanding, if i miss something, please let me know and i'll add it. thanks for big helping!

**Comment by Usama3059** [comment]
Hi @glenn-jocher  Thanks for the amazing tips I have a problem since I'm training on a large dataset around 15k on Google Colab
Training doesn't start as it ends on close to 30% of creating a cache file for training images.

**Comment by glenn-jocher** [comment]
@Usama3059 you can `--cache` your dataset **if your hardware has sufficient RAM**. If your system has insufficient RAM then clearly don't use this option.
